cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/runtime/KubeTwillRunnerService.java

```diff
@@ -252,6 +316,378 @@ public void start() {
     }
   }
 
+  @Override
+  public void onStart(Collection<NamespaceDetail> namespaceDetails) {
+    for (NamespaceDetail namespaceDetail : namespaceDetails) {
+      String namespace = namespaceDetail.getProperties().get(KubeMasterEnvironment.NAMESPACE_PROPERTY);
+      // Start job watchers on non-system k8s namespaces. We add watchers on the system namespace in start()
+      if (namespace != null && !resourceWatchers.containsKey(namespace) && !namespace.equals(kubeNamespace)) {
+        addAndStartJobWatcher(namespace);
+      }
+    }
+  }
+
+  @Override
+  public void onNamespaceCreation(NamespaceDetail namespaceDetail) throws Exception {
+    String cdapNamespace = namespaceDetail.getName();
+    Map<String, String> properties = namespaceDetail.getProperties();
+    if (NamespaceId.isReserved(cdapNamespace)) {
+      return;
+    }
+
+    String namespace = properties.get(KubeMasterEnvironment.NAMESPACE_PROPERTY);
+    if (namespace == null || namespace.isEmpty()) {
+      throw new IOException(String.format("Cannot create Kubernetes namespace for %s because no name was provided",
+                                          cdapNamespace));
+    }
+    // Kubernetes namespace must be a lowercase RFC 1123 label, consisting of lower case alphanumeric characters or '-'
+    // and must start and end with an alphanumeric character
+    KubeUtil.validateRFC1123LabelName(namespace);
+    findOrCreateKubeNamespace(namespace, cdapNamespace);
+    updateOrCreateResourceQuota(namespace, cdapNamespace, properties);
+    copyVolumes(namespace, cdapNamespace);
+    createWorkloadServiceAccount(namespace, cdapNamespace);
+    if (workloadIdentityEnabled) {
+      String workloadIdentityServiceAccountEmail = properties.get(WORKLOAD_IDENTITY_GCP_SERVICE_ACCOUNT_EMAIL_PROPERTY);
+      if (workloadIdentityServiceAccountEmail != null && !workloadIdentityServiceAccountEmail.isEmpty()) {
+        findOrCreateWorkloadIdentityConfigMap(namespace, workloadIdentityServiceAccountEmail);
+      }
+    }
+    addAndStartJobWatcher(cdapNamespace);
+  }
+
+  @Override
+  public void onNamespaceDeletion(NamespaceDetail namespaceDetail) throws Exception {
+    String namespace = namespaceDetail.getProperties().get(KubeMasterEnvironment.NAMESPACE_PROPERTY);
```

Comment:
seems like we should also stop and remove the watchers for the namespace

Reply:
added stopJobWatcher()